src/qml_benchmarks/data/ising.py

```diff
@@ -0,0 +1,234 @@
+    """
+    def __init__(
+        self, N: int, J: jnp.array, b: jnp.array, T: float, sparse=False, compute_partition_fn=False
+    ) -> None:
+
+        self.N = N
+        self.kernel = MetropolisHastings()
+        self.J = J
+        self.T = T
+        self.b = b
+        self.J_sparse = jnp.nonzero(J) if sparse else None
+
+        if compute_partition_fn:
+            Z = 0
+            for i in tqdm(range(2**self.N), desc="Computing partition function"):
+                lattice = (-1) ** jnp.array(jnp.unravel_index(i, [2] * self.N))
+                en = energy(lattice, self.J, self.b, self.J_sparse)
+                Z += jnp.exp(-en / T)
+            self.Z = Z
+
+    def sample(
+        self, num_samples: int, num_chains=1, thinning=1, num_warmup=1000, key=42
+    ) -> jnp.array:
+
+        """
+        Generate samples.
+        Args:
+            num_samples (int): total number of samples to generate per chain
+            num_chains (int): number of chains
+            thinning (int): how much to thin the sampling. e.g. if thinning = 10 a sample will be drawn ater each
+                10 steps of mcmc sampling. Larger numbers result in more unbiased samples.
+            num_warmup (int): number of mcmc 'burn in' steps to perform before collecting any samples.
+            key (int): random seed used to initialize sampling.
+        """
+        rng_key = random.PRNGKey(key)
+        init_spins = initialize_spins(rng_key, self.N, num_chains)
+        mcmc = MCMC(
+            self.kernel,
+            num_warmup=num_warmup,
+            thinning=thinning,
+            num_samples=num_samples,
+            num_chains=num_chains,
+        )
+        mcmc.run(
+            rng_key,
+            init_params=init_spins,
+            J=self.J,
+            b=self.b,
+            T=self.T,
+            J_sparse=self.J_sparse,
+        )
+        samples = mcmc.get_samples()
+        samples.reshape((-1, self.N))
+        return (samples+1)//2
+
+    def probability(self, x: ndarray) -> float:
+        """
+        compute the probability of a binary configuration x
+        Args:
+            x: binary configuration array
+        Returns:
+            (float): the probability of sampling x according to the ising distribution
+        """
+
+        if not(hasattr(self, 'Z')):
+            raise Exception('probability requires partition fuction to have been computed')
+
+        return (
+            jnp.exp(-energy(x, self.J, self.b, self.J_sparse) / self.T)
+            / self.Z
+        )
+
+def generate_ising(N: int,
+                   num_samples: int,
+                   J: jnp.array,
+                   b: jnp.array,
+                   T: float,
+                   sparse=False,
+                   num_chains=1,
+                   thinning=1,
+                   num_warmup=1000,
+                   key=42):
+    r"""
+    Generating function for ising datasets.
```

Comment:
```suggestion
    Generating function for Ising datasets.
```
Minor, but looks more professional to use the right capitalisation?